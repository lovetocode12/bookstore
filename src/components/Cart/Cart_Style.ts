import styled from "styled-components";

export const ContainerDiv = styled.div`
    width: 100%;
    display:flex;
    justify-content: space-between; 
    flex-wrap:wrap;
    
    .buttonsContainer{
        display: flex;
        justify-content: space-evenly;
        margin: 1em;
        flex-wrap: wrap;

    }
    .buttonsContainer>button{
        margin: 1em;
        color: white;
        background: #ea8686;
        padding: 0.5em 5.5em;
        font-size: 0.7rem;
        cursor: pointer;
        border-radius: 5px;
        border: none;
    }
    .title{
        margin: 0.5em 1em;
        font-size: 1.3em;
    }
    `

export const AdressDiv = styled.div`
     margin: 1em;
    flex-grow: 1;
    flex-basis: 20em;
    height: 50%;
   `


export const InputsDiv = styled.div<{ submitted: boolean }>`
height: 100%;
display:flex;
flex-direction: column;
flex-wrap:wrap;
background-color: #e2c7c73b;

.div_input_conatiner{
    margin: 0.6em 0;
    padding: 0.3em 1em;
    display: flex;
    justify-content: space-between;
    flex-wrap:wrap;
}

.div_input_conatiner>label{
    margin-right:1em;
    min-width: 10em;
}
.div_input_conatiner>input{
    width: 24em;
    pointer-events: ${props => props.submitted ? 'none' : 'initial'};
    background-color: ${props => props.submitted ? '#f5e5e53b !important' : 'white'};
    border: ${props => props.submitted ? 'none' : '2px solid #bcb4b4'};
    padding: 0.3em;
}
`


export const CheckoutDiv = styled.div`
    flex-grow: 1;
    margin: 1em;
    flex-basis: 20em;  
    height:100%;

    .booksContainer{
        display:flex;
        flex-direction:column;
        background-color: #e2c7c73b;
        min-height: 12em;
         overflow: auto;
    }

    .booksContainer>ul{
        list-style:none;
    }


    .order_item{
        display:flex;
        justify-content: space-between;
        margin: 1em;
        font-size: 0.9em;
    }

    .paymentContainer{
        flex-grow:2;
    }
    `