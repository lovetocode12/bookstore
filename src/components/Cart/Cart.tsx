import React, { useState, useEffect, useContext } from 'react'
import { CheckoutDiv, AdressDiv, ContainerDiv, InputsDiv } from './Cart_Style'
import { useFormik } from "formik";
import { useSelector, useDispatch } from 'react-redux';
import { Address } from '../../store/orders/type';
import { checkOut, clearCart, setAddress } from '../../store/orders/order_action';
import { TitleContext } from '../../common/TitleContext';
import { RootState } from '../../store/type';


function Cart() {

    const [submitted, setSubmitted] = useState(false);
    const dispatch = useDispatch();
    const { orders, address } = useSelector((state: RootState) => {
        return { orders: state.orderState.currentOrders, address: state.orderState.address }
    });
    const context:any = useContext(TitleContext);
    const dispatchTitle = context.dispatchTitle;
    useEffect(() => {
        if (dispatchTitle) {
            dispatchTitle({ type: 'SET_TITLE', payload: 'CART' })
        }
    }, [dispatchTitle])

    const taxPercent = 5;
    let totalPrice = 0;
    let shippingCharge = 0;
    let taxAmount = 0;
    let itemsPrice = orders?.length && orders.reduce((acc, order) => acc + order.book.price, 0)

    if (itemsPrice > 0) {
        shippingCharge = orders.length * 3;
        taxAmount = itemsPrice * (taxPercent / 100);
        totalPrice = itemsPrice + shippingCharge + taxAmount;
    }

    const formik = useFormik({
        initialValues: {
            address1: address?.address1 || '',
            address2: address?.address2 || '',
            email: address?.email || '',
            phone: address?.phone || '',
            city: address?.city || ''
        },
        onSubmit: (values: Address) => {
            setSubmitted(state => !state);
            dispatch(setAddress(values))
            // alert(JSON.stringify(values, null, 2));
        },
    });
    return (
        <ContainerDiv>
            <AdressDiv>
                <div className="title">{"Shipping Address"}</div>
                <form onSubmit={formik.handleSubmit}>
                    <InputsDiv submitted={submitted}>
                        <div className="div_input_conatiner">
                            <label htmlFor="address1">Address 1</label>
                            <input
                                id="address1"
                                name="address1"
                                type="text"
                                onChange={formik.handleChange}
                                value={formik.values.address1}
                            />
                        </div>
                        <div className="div_input_conatiner"><label htmlFor="address2">Address 2</label>
                            <input
                                id="address2"
                                name="address2"
                                type="text"
                                onChange={formik.handleChange}
                                value={formik.values.address2}
                            /></div>
                        <div className="div_input_conatiner"><label htmlFor="email">Email</label>
                            <input
                                id="email"
                                name="email"
                                type="text"
                                onChange={formik.handleChange}
                                value={formik.values.email}
                            /></div>
                        <div className="div_input_conatiner"><label htmlFor="phone">Phone Number</label>
                            <input
                                id="phone"
                                name="phone"
                                type="text"
                                onChange={formik.handleChange}
                                value={formik.values.phone}
                            /></div>
                        <div className="div_input_conatiner"><label htmlFor="city">City</label>
                            <input
                                id="city"
                                name="city"
                                type="text"
                                onChange={formik.handleChange}
                                value={formik.values.city}
                            /></div>
                    </InputsDiv>
                    <div className="buttonsContainer">
                        <button type="submit">Save Address</button>
                        <button type="submit">Edit Address</button>
                    </div>
                </form>
            </AdressDiv>
            <CheckoutDiv >
            <div className="title">{"Shopping Bag"}</div>
                <div className="booksContainer">
                    <ul>
                        {
                            orders?.length > 0 ? orders.map((order, index) =>
                                <li className="order_item" key={index}>
                                    <div>{order.book.title}</div>
                                    <div className="price">{order.book.price}</div>
                                </li>) : 'No Items In the Cart'
                        }
                    </ul>
                </div>
                <div className="paymentContainer">
                    <div className="mediumHeading">Payment Info</div>
                    <div className="col-2">
                        <div>Items Price</div>
                        <div className="price">{itemsPrice}</div>
                    </div>
                    <div className="col-2">
                        <div>Tax</div>
                        <div className="price">{taxAmount}</div>
                    </div>
                    <div className="col-2">
                        <div>Shipping Charges</div>
                        <div className="price">{shippingCharge}</div>
                    </div>
                    <hr />
                    <div className="col-2">
                        <div>Total</div>
                        <div className="price">{totalPrice}</div>
                    </div>
                    <div className="buttonsContainer">
                        <button type="button" onClick={() => dispatch(checkOut())}>Checkout</button>
                        <button type="button" onClick={() => dispatch(clearCart())}>Cancel</button>
                    </div>
                </div>
            </CheckoutDiv>
        </ContainerDiv>
    )
}

export default Cart
