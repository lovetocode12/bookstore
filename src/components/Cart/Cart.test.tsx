import React from 'react';
import { shallow } from '../../../enzyme';
import Cart from './Cart';
import * as ReactReduxHooks from 'react-redux';

const spyDispatch = jest
    .spyOn(ReactReduxHooks, "useDispatch")
    .mockImplementation(() => jest.fn());
jest
    .spyOn(ReactReduxHooks, "useSelector")
    .mockImplementation(() => jest.fn());

describe('CART Component', () => {

    function getWrapper() {
        const wrapper: any = shallow(<Cart />);
        return wrapper;
    }

    it('Component should render correctly', () => {
        const wrapper = getWrapper();
        expect(wrapper).toMatchSnapshot();
    })

    it('Component should change all input values correctly', () => {
        const wrapper = getWrapper();
        const inputAddress1 = wrapper.find('#address1')
        const inputAddress2 = wrapper.find('#address2')
        const inputPhone = wrapper.find('#phone')
        const inputEmail = wrapper.find('#email')
        const inputCity = wrapper.find('#city')
        inputAddress1.value = 'line 1';
        inputAddress2.value = 'line 2';
        inputPhone.value = '12345';
        inputEmail.value = 'test@c.com';
        inputCity.value = 'hyd';
        wrapper.update();
        expect(inputAddress1.value).toEqual('line 1');
        expect(inputAddress2.value).toEqual('line 2');
        expect(inputPhone.value).toEqual('12345');
        expect(inputEmail.value).toEqual('test@c.com');
        expect(inputCity.value).toEqual('hyd');
        const submitButton = wrapper.find('form');
        submitButton.simulate('submit');
        expect(spyDispatch).toHaveBeenCalled();
    })

});