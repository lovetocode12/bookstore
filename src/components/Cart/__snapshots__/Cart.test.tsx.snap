// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CART Component Component should render correctly 1`] = `
<styled.div>
  <styled.div>
    <div
      className="title"
    >
      Shipping Address
    </div>
    <form
      onSubmit={[Function]}
    >
      <styled.div
        submitted={false}
      >
        <div
          className="div_input_conatiner"
        >
          <label
            htmlFor="address1"
          >
            Address 1
          </label>
          <input
            id="address1"
            name="address1"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="div_input_conatiner"
        >
          <label
            htmlFor="address2"
          >
            Address 2
          </label>
          <input
            id="address2"
            name="address2"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="div_input_conatiner"
        >
          <label
            htmlFor="email"
          >
            Email
          </label>
          <input
            id="email"
            name="email"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="div_input_conatiner"
        >
          <label
            htmlFor="phone"
          >
            Phone Number
          </label>
          <input
            id="phone"
            name="phone"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="div_input_conatiner"
        >
          <label
            htmlFor="city"
          >
            City
          </label>
          <input
            id="city"
            name="city"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </styled.div>
      <div
        className="buttonsContainer"
      >
        <button
          type="submit"
        >
          Save Address
        </button>
        <button
          type="submit"
        >
          Edit Address
        </button>
      </div>
    </form>
  </styled.div>
  <styled.div>
    <div
      className="title"
    >
      Shopping Bag
    </div>
    <div
      className="booksContainer"
    >
      <ul>
        No Items In the Cart
      </ul>
    </div>
    <div
      className="paymentContainer"
    >
      <div
        className="mediumHeading"
      >
        Payment Info
      </div>
      <div
        className="col-2"
      >
        <div>
          Items Price
        </div>
        <div
          className="price"
        />
      </div>
      <div
        className="col-2"
      >
        <div>
          Tax
        </div>
        <div
          className="price"
        >
          0
        </div>
      </div>
      <div
        className="col-2"
      >
        <div>
          Shipping Charges
        </div>
        <div
          className="price"
        >
          0
        </div>
      </div>
      <hr />
      <div
        className="col-2"
      >
        <div>
          Total
        </div>
        <div
          className="price"
        >
          0
        </div>
      </div>
      <div
        className="buttonsContainer"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          Checkout
        </button>
        <button
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </styled.div>
</styled.div>
`;
